Composer/packages/adaptive-form/src/components/fields/RecognizerField/useMigrationEffect.ts

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { useShellApi, ChangeHandler } from '@bfc/extension-client';
+import { useEffect } from 'react';
+import { MicrosoftIRecognizer } from '@bfc/shared';
+
+export const useMigrationEffect = (
+  recognizer: MicrosoftIRecognizer | undefined,
+  onChangeRecognizer: ChangeHandler<MicrosoftIRecognizer>
+) => {
+  const { qnaFiles, luFiles, currentDialog, locale } = useShellApi();
+
+  useEffect(() => {
+    // this logic is for handling old bot with `recognizer = undefined'
+    if (recognizer === undefined) {
+      const qnaFile = qnaFiles.find((f) => f.id === `${currentDialog.id}.${locale}`);
+      const luFile = luFiles.find((f) => f.id === `${currentDialog.id}.${locale}`);
+      if (qnaFile && luFile) {
+        onChangeRecognizer(`${currentDialog.id}.lu.qna`);
+      }
+    }
+
+    // transform lu recognizer to crosstrained for old bot
+    if (recognizer === `${currentDialog.id}.lu`) {
```

Comment:
@liweitian How do you consider the migration logic when authoring the QnA PR?

Reply:
This is for migrating old bot whose recognizer property is `${currentDialog.id}.lu`. transform it for `luis & qna` recognizer

Reply:
@alanlong9278 Will it affect users whose runtime version is older than the version with QnA?
Those outdated runtimes are still using "*.lu" files as recognizer.

Reply:
To my understanding, for old bots there are 2 cases:
1. Old bots **without** an ejected runtime - No problem since they use Composer's runtime template which is always updated
2. Old bots **with** ejected runtime - Migration to qna will break them since QnA is not supported in their ejected runtime.

Reply:
@a-b-r-o-w-n I think it's worthy for us to open a new issue for tracking this, the migration effect do break some old bots.

Reply:
Ok. Did someone on your team already create an issue?

Reply:
Just created one https://github.com/microsoft/BotFramework-Composer/issues/4229